### ISSUE TITLE
Which one is suitable for gas retrieval?

### ISSUE THREAD

**Issue by zxdawn** [Helmholtz-AI-Energy/HyDe#23]
Hi, it's nice to see so many Denoising algorithms for HSI. May I ask which one is suitable for retrieval? I have tried `HyRes` and `FastHyDe` with default settings, and the denoised result of each band looks great. However, when I try to pick the signals using the matched filter, I can't see anything. Maybe, the signals are removed in the denoising steps.

Here's the difference of data for one band. It seems the differences are large for "not noisy" regions.

<img width="1005" alt="image" src="https://github.com/Helmholtz-AI-Energy/HyDe/assets/30388627/c67ba09c-29fa-4068-ad29-324172196abb">

**Comment by coquelin77** [comment]
Hello, thank you for asking the question, I hope you are finding the package easy to use. 

All of the methods should be suitable for retrieval (as far as I know). Although, I must admit that I am not an expert in what you are attempting to do. 

It should be noted that HyRes and FastHyDe are both intended to remove normal Gaussian noise (commonly referred to as 'blurring'), I am not sure if that helps.

I have noticed during some tests that many methods perform better when the images are normalized:
```python
input_tens, norms = hyde.normalize(input_tens, by_band=False)
...
# operate on normalized image
...
output = hyde.undo_normalize(output, **norms)
```
This might help you get better results. Judging by the difference that you have shown in the rightmost image, it looks like the denoising method is not doing much at all. 

I hope this helps, but im not sure if i can help without knowing more about the use-case.

please let me know if this helps and/or you have more questions!

**Comment by zxdawn** [comment]
Hi Daniel, thank you for your detailed explanation. It's convenient to apply these methods by HyDe.

I checked some data which makes this problem more obvious. My 3D data has 127 bands and there're some stripes and noises. Let me take No. 88 as an example:

![image](https://github.com/Helmholtz-AI-Energy/HyDe/assets/30388627/4c968edf-389a-4279-aaba-f57353ff6239)

You can see some vertical stripes there. Then, I tested `HyRes` by subsetting data with different band numbers (all 127, 39, and 10 bands).

Here're the results:

![image](https://github.com/Helmholtz-AI-Energy/HyDe/assets/30388627/ed5c581e-fc5c-4c96-838b-f13133f673da)

![image](https://github.com/Helmholtz-AI-Energy/HyDe/assets/30388627/3927bc2d-6fd7-4825-a848-286eb9bb5100)

![image](https://github.com/Helmholtz-AI-Energy/HyDe/assets/30388627/dde936d7-c22c-44a0-9ed9-12ab5fa6ce2f)

The red pixels in the first figure (127 bands) are the signals I wanna keep. It seems more bands are included, more signals are removed. I suppose that's because the strength of signals is different for different bands.

Is it possible to remove background noise, but keep the signals?

In case you wanna check the data, here's the [npy file](https://drive.google.com/file/d/1BPINVcrX8VDr62BApgpFF0Xc9PtTT-jY/view?usp=sharing) (~ 440 MB) and the simple script to reproduce the first figure:

```
import numpy as np
import matplotlib.pyplot as plt

data = np.load('test.npy')
input_tens = torch.tensor(data, dtype=torch.float32, device="cpu")
input_tens, norms = hyde.normalize(input_tens, by_band=False)
hyres = hyde.HyRes()
output = hyres(input_tens)
output = hyde.undo_normalize(output, **norms)

index_band = 88


fig, axs = plt.subplots(figsize=(16, 8), ncols=3)

m = axs[0].imshow(data[:, :, index_band], cmap='viridis', vmin=4.2, vmax=6)
plt.colorbar(m, ax=axs[0], extend='both', shrink=0.5)

m = axs[1].imshow(output[:, :, index_band], cmap='viridis', vmin=4.2, vmax=6)
plt.colorbar(m, ax=axs[1], extend='both', shrink=0.5)

m = axs[2].imshow((output-data)[:, :, index_band], cmap='RdBu_r', vmin=-0.5, vmax=0.5)
plt.colorbar(m, ax=axs[2], extend='both', shrink=0.5)

axs[0].set_title('Original')
axs[1].set_title('Denoise')
axs[2].set_title('Denoise - Original')

plt.suptitle(f'Number of total bands: {data.shape[-1]}', y=0.8)
plt.tight_layout()
```

Thanks for your kind help again!
Xin

**Comment by coquelin77** [comment]
It took me some time to see this, but I think I have found the issue: there are strips in the original, which are being removed in the denoised image (although this is not something that HyRes is intended for). So HyRes is attempting to take those bands out, which is causing the stripping pattern in the difference plot you showed. You can see if you compare column 808. 

Does this help?

**Comment by zxdawn** [comment]
Ha, thanks. I thought HyRes can also remove stripes besides noises. Anyway, I will try to remove the stripes using Fourier transform and then denoise the data again. Thank you again and close the issue now.
